lib/code-caller-docker.js

```diff
@@ -0,0 +1,783 @@
+        this._checkState();
+        this.debug('exit call()');
+      })
+      .catch((err) => callback(err));
+  }
+
+  /**
+   * @param {CallType} type
+   * @param {string} directory
+   * @param {string} file
+   * @param {string} fcn
+   * @param {any[]} args
+   * @returns {Promise<{ result: any, output: string }>}
+   */
+  callAsync(type, directory, file, fcn, args) {
+    return new Promise((resolve, reject) => {
+      this.call(type, directory, file, fcn, args, (err, data, output) => {
+        if (err) return reject(err);
+        return resolve({ result: data, output });
+      });
+    });
+  }
+
+  /**
+   *
+   * @param {(err: Error | null | undefined, success?: boolean) => void} callback
+   */
+  restart(callback) {
+    util.callbackify(this.restartAsync)(callback);
+  }
+
+  async restartAsync() {
+    this.debug('enter restart()');
+    if (!this._checkState([CREATED, WAITING, EXITING, EXITED])) {
+      throw new Error('Unexpected DockerCaller state');
+    }
+
+    if (this.callCount === 0) {
+      // If there weren't any calls since the last time this code caller
+      // was restarted, we can slightly optimize things by skipping the
+      // restart. This is safe, as no user-provided code will have been
+      // loaded into the Python interpreter.
+      this.debug(`exit restart() - skipping since no calls recorded since last restart`);
+      return true;
+    } else if (this.state === CREATED) {
+      // no need to restart if we don't have a worker
+      this.debug(`exit restart()`);
+      return true;
+    } else if (this.state === WAITING) {
+      const { result } = await this.callAsync('restart', null, null, 'restart', []);
+      this.coursePath = null;
+      this.callCount = 0;
+      if (result !== 'success') throw new Error(`Error while restarting: ${result}`);
+      this.debug('exit restart()');
+      return true;
+    } else if (this.state === EXITING || this.state === EXITED) {
+      this.debug('exit restart()');
+      return false;
+    }
+  }
+
+  done() {
+    this.debug('enter done()');
+    this._checkState([CREATED, WAITING, EXITING, EXITED]);
+
+    if (this.state === CREATED) {
+      this.state = EXITED;
+    } else if (this.state === WAITING) {
+      this._cleanup();
+      this.state = EXITING;
+    }
+    this._checkState();
+    this.debug('exit done()');
+  }
+
+  async ensureChild() {
+    this.debug('enter ensureChild()');
+    this._checkState();
+
+    // Since container creation is async, it's possible that ensureChild()
+    // could be called again while it's already executing. For instance, we
+    // could be inside a call that was made to warm up this worker, but then
+    // we might have call() invoked, which will also call ensureChild(). We
+    // need to ensure that we're only ever inside this function once at a time,
+    // so we'll use a "mutex" (even though we're in a single-threaded environment).
+    await this.ensureChildMutex.runExclusive(async () => {
```

Comment:
Rather than using a mutex here, would it be cleaner to handle this via the state machine with a STARTING state? I'm slightly nervous about expanding our set of states with these extra mutex states --- our full state is now the tuple `(this.state, ensureChildMutex, ensureDirectoryMutex)` which feels harder to reason about.

Reply:
IMO it's actually *easier* to reason about with a mutex. With a mutex, we can just blindly call `prepareForCourse` whenever we want and be assured that for any caller, by the time it returns, we're prepared. Without a mutex, we'd need some way for one call to react to the first call completing - event emitters? More callbacks? Either way, I think the other options are more complex than just putting a mutex here. We can talk this through on Zoom tomorrow though.

Reply:
We decided that using mutexes is good here, but we could get away with only one of them: https://github.com/PrairieLearn/PrairieLearn/pull/1626#discussion_r923943058